File: tvr-decoder/src/main/java/com/willcurrie/decoders/ICCPublicKeyDecoder.java
package com.willcurrie.decoders;

import com.willcurrie.tlv.ISOUtil;

public class ICCPublicKeyDecoder {

    public String decode(byte[] recovered, int byteLengthOfIssuerModulus) {
        StringBuilder b = new StringBuilder();
        b.append("Header: ").append(ISOUtil.hexString(recovered, 0, 1)).append('\n');
        b.append("Format: ").append(ISOUtil.hexString(recovered, 1, 1)).append('\n');
        b.append("PAN: ").append(ISOUtil.hexString(recovered, 2, 10)).append('\n');
        b.append("Expiry Date (MMYY): ").append(ISOUtil.hexString(recovered, 12, 2)).append('\n');
        b.append("Serial number: ").append(ISOUtil.hexString(recovered, 14, 3)).append('\n');
        b.append("Hash algorithm: ").append(ISOUtil.hexString(recovered, 17, 1)).append('\n');
        b.append("Public key algorithm: ").append(ISOUtil.hexString(recovered, 18, 1)).append('\n');
        String publicKeyLength = ISOUtil.hexString(recovered, 19, 1);
        int i = Integer.parseInt(publicKeyLength, 16);
        b.append("Public key length: ").append(publicKeyLength).append(" (").append(i).append(")").append('\n');
        b.append("Public key exponent length: ").append(ISOUtil.hexString(recovered, 20, 1)).append('\n');
        b.append("Public key: ").append(ISOUtil.hexString(recovered, 21, byteLengthOfIssuerModulus-42)).append('\n');
//        b.append("          : ").append(ISOUtil.hexString(recovered, 20, i)).append('\n');
        b.append("Hash: ").append(ISOUtil.hexString(recovered, 21 + byteLengthOfIssuerModulus-42, 20)).append('\n');
        b.append("Trailer: ").append(ISOUtil.hexString(recovered, 21+byteLengthOfIssuerModulus-42+20, 1)).append('\n');
        return b.toString();
    }
}

File: src/main/java/io/github/binaryfoo/decoders/ICCPublicKeyDecoder.kt
package io.github.binaryfoo.decoders

import io.github.binaryfoo.tlv.ISOUtil
import io.github.binaryfoo.decoders.annotator.SignedDataDecoder
import io.github.binaryfoo.EmvTags
import io.github.binaryfoo.crypto.RecoveredPublicKeyCertificate
import io.github.binaryfoo.DecodedData
import io.github.binaryfoo.findAllForTag
import io.github.binaryfoo.findTlvForTag
import io.github.binaryfoo.findValueForTag

/**
 * EMV 4.3 Book2, Table 14: Format of Data Recovered from ICC Public Key Certificate
 */
public class ICCPublicKeyDecoder : SignedDataDecoder {

    override fun decodeSignedData(session: DecodeSession, decoded: List<DecodedData>) {
        val recoveredIssuerPublicKeyCertificate = session.issuerPublicKeyCertificate
        val iccCertificate = decoded.findTlvForTag(EmvTags.ICC_PUBLIC_KEY_CERTIFICATE)

        if (iccCertificate != null && recoveredIssuerPublicKeyCertificate != null && recoveredIssuerPublicKeyCertificate.exponent != null) {
            for (decodedCertificate in decoded.findAllForTag(EmvTags.ICC_PUBLIC_KEY_CERTIFICATE)) {
                val result = recoverCertificate(iccCertificate, decodedCertificate, recoveredIssuerPublicKeyCertificate, ::decodeICCPublicKeyCertificate)
                if (result.certificate != null) {
                    result.certificate.rightKeyPart = decoded.findValueForTag(EmvTags.ICC_PUBLIC_KEY_REMAINDER)
                    result.certificate.exponent = decoded.findValueForTag(EmvTags.ICC_PUBLIC_KEY_EXPONENT)
                    session.iccPublicKeyCertificate = result.certificate
                }
            }
        }
    }

}

fun decodeICCPublicKeyCertificate(recovered: ByteArray, byteLengthOfIssuerModulus: Int, startIndexInBytes: Int): RecoveredPublicKeyCertificate {
    val publicKeyLength = Integer.parseInt(ISOUtil.hexString(recovered, 19, 1), 16)
    val exponentLength = ISOUtil.hexString(recovered, 20, 1)
    var lengthOfLeftKeyPart = if (publicKeyLength > byteLengthOfIssuerModulus - 42) byteLengthOfIssuerModulus - 42 else publicKeyLength
    val leftKeyPart = ISOUtil.hexString(recovered, 21, lengthOfLeftKeyPart)
    val children = listOf(
        DecodedData.byteRange("Header", recovered, 0, 1, startIndexInBytes),
        DecodedData.byteRange("Format", recovered, 1, 1, startIndexInBytes),
        DecodedData.byteRange("PAN", recovered, 2, 10, startIndexInBytes),
        DecodedData.byteRange("Expiry Date (MMYY)", recovered, 12, 2, startIndexInBytes),
        DecodedData.byteRange("Serial number", recovered, 14, 3, startIndexInBytes),
        DecodedData.byteRange("Hash algorithm", recovered, 17, 1, startIndexInBytes),
        DecodedData.byteRange("Public key algorithm", recovered, 18, 1, startIndexInBytes),
        DecodedData.byteRange("Public key length", publicKeyLength.toString(), 19, 1, startIndexInBytes),
        DecodedData.byteRange("Public key exponent length", exponentLength, 20, 1, startIndexInBytes),
        DecodedData.byteRange("Public key", leftKeyPart, 21, lengthOfLeftKeyPart, startIndexInBytes),
        DecodedData.byteRange("Hash", recovered, 21 + byteLengthOfIssuerModulus - 42, 20, startIndexInBytes),
        DecodedData.byteRange("Trailer", recovered, 21 + byteLengthOfIssuerModulus - 42 + 20, 1, startIndexInBytes)
    )
    return RecoveredPublicKeyCertificate("ICC", children, exponentLength, leftKeyPart)
}
